### PR TITLE
Implement setGraph, refactor Node interface

### PR DESCRIPTION
**Pull request by prybicki**
- Introduce `IStreamBound` to allow changing stream bound to the `DeviceArrayAsync` instance
- Introduce `DeviceAsyncArrayManager` [I'd appreciate a shorter name idea] to conveniently change the stream of a group of DAAs (e.g. all used by a node)
- Refactor Node methods:
    - `onInputChange()` -> `validateImpl()`
    - `schedule()` -> `execEnqueueImpl()`
    - Add their non-virtual trampolines to allow executing code before and after calling the virtual override of the derived class
    - Rename other helper methods to improve readability
- Remove some too-invasive tests of `Array` (counting memory operations)
    - They are too costly to maintain vs. benefit from having them
- Revise addChild/removeChild
- Add tests demonstrating that DeviceArrayAsync's stream can be changed, either directly or via the manager
- Add tests demonstrating adding/removing the child behaves correctly, even when the graph is changed
- Update addChild/removeChild
    - Set dirty flag whenever graph structure is changed

### CODE REVIEW

File: src/memory/DeviceAsyncArrayManager.hpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		if (lastStream.expired()) {
+			auto msg = fmt::format("DeviceAsyncArrayManager: called create() without prior call to setStream()");
+			throw std::logic_error(msg);
+		}
+		auto ptr = DeviceAsyncArray<T>::create(lastStream.lock());
+		streamBoundObjects.emplace_back(ptr);
+		return ptr;
+	}
+
+	void setStream(CudaStream::Ptr newStream)
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/memory/DeviceAsyncArrayManager.hpp:38]: Rejected offline

File: test/src/ArrayTest.cpp [hunk 2]
```diff
@@ -32,77 +33,87 @@ TEST(Array, Typing)
...
+        arrayMgr.setStream(streamA);
+        DeviceAsyncArray<int>::Ptr array = arrayMgr.create<int>();
 
-struct ArrayOps : public ::testing::Test
+        array->resize(5, true, true);
+        CHECK_CUDA(cudaStreamSynchronize(streamA->get()));
+
+        // Destroy streamA so it is unusable.
+        // It will cause an error meessage on second attempt to destroy, but it does not affect the test.
+        CHECK_CUDA(cudaStreamDestroy(streamA->get()));
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/ArrayTest.cpp:63]: I will use different mechanism to implement this test.

File: test/src/ArrayTest.cpp [hunk 3]
```diff
@@ -32,77 +33,87 @@ TEST(Array, Typing)
...
+        EXPECT_THROW(array->resize(10, true, true), std::runtime_error);
+
+        // After setting a new stream, resize should work.
+        arrayMgr.setStream(streamB);
+        array->resize(10, true, true);
+        CHECK_CUDA(cudaStreamSynchronize(streamB->get()));
+}
+
+TEST(Array, ChangeStreamViaManager)
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/ArrayTest.cpp:72]: ACK

File: src/memory/ConcreteArrays.hpp [hunk 4]
```diff
@@ -116,6 +137,13 @@ struct HostPinnedArray : public HostArray<MemoryKind::HostPinned, T>
 		};
 	}
 
+	void copyFrom(DeviceAsyncArray<T>::Ptr source)
+	{
+		this->resize(source->getCount(), false, false);
+		CHECK_CUDA(cudaMemcpyAsync(this->data, source->data, this->getSizeOf() * this->getCount(), cudaMemcpyDeviceToHost, source->getStream()->get()));
```
Comment (PiotrMrozik) [inline src/memory/ConcreteArrays.hpp:143]: I think this copying is wrong.

File: src/memory/StreamBoundObjectsManager.hpp [hunk 5]
```diff
@@ -17,27 +17,35 @@
...
-	DeviceAsyncArray<T>::Ptr create()
+	void registerObject(IStreamBound::Ptr object)
 	{
-		if (lastStream.expired()) {
-			auto msg = fmt::format("DeviceAsyncArrayManager: called create() without prior call to setStream()");
+		if (!currentStream.has_value()) {
+			currentStream = object->getStream();
+		}
+		if (currentStream.value().expired()) {
+			auto msg = fmt::format("StreamBoundObjectsManager: called registerObject(), but current stream has expired");
```
Comment (PiotrMrozik) [inline src/memory/StreamBoundObjectsManager.hpp:32]: In that case shouldn't we take object stream?

**Review by PiotrMrozik — APPROVED** [review]

**Pull request merged**